### ISSUE TITLE
InvisibleCaptcha::ControllerExt honeypot attribute ActiveSupport::Notifications dispatches an `invisible_captcha.spam_detected` event test is failing now

### ISSUE THREAD

**Issue by pravi** [markets/invisible_captcha#118]
```
$ bundle install
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Using rake 13.0.6
Using concurrent-ruby 1.2.2
Using minitest 5.17.0
Using builder 3.2.4
Using racc 1.6.2
Using nio4r 2.5.8
Using websocket-extensions 0.1.5
Using marcel 1.0.2
Using rack 2.2.6.2
Using mini_mime 1.1.2
Using date 3.3.3
Using timeout 0.3.2
Using bundler 2.3.15
Using thor 1.2.1
Using diff-lcs 1.5.0
Using method_source 1.0.0
Using zeitwerk 2.6.7
Using rspec-support 3.12.0
Using i18n 1.12.0
Using tzinfo 2.0.6
Using nokogiri 1.14.2 (x86_64-linux)
Using websocket-driver 0.7.5
Using crass 1.0.6
Using activesupport 7.0.4.2
Using rack-test 2.0.2
Using appraisal 2.4.1
Using rspec-core 3.12.1
Using rspec-expectations 3.12.2
Using rspec-mocks 3.12.3
Using net-protocol 0.2.1
Using erubi 1.12.0
Using globalid 1.1.0
Using loofah 2.19.1
Fetching sprockets 4.2.0
Using rails-dom-testing 2.0.3
Using activemodel 7.0.4.2
Using net-imap 0.3.4
Using net-smtp 0.3.3
Using net-pop 0.1.2
Using activejob 7.0.4.2
Using rails-html-sanitizer 1.5.0
Using actionview 7.0.4.2
Using activerecord 7.0.4.2
Using mail 2.8.1
Using actionpack 7.0.4.2
Using actioncable 7.0.4.2
Using activestorage 7.0.4.2
Using actionmailer 7.0.4.2
Using actiontext 7.0.4.2
Using railties 7.0.4.2
Using actionmailbox 7.0.4.2
Using rspec-rails 6.0.1
Using rails 7.0.4.2
Using invisible_captcha 2.0.0 from source at `.`
Installing sprockets 4.2.0
Fetching sprockets-rails 3.4.2
Installing sprockets-rails 3.4.2
Bundle complete! 4 Gemfile dependencies, 55 gems now installed.
Bundled gems are installed into `./vendor/bundle`
$ bundle exec rake spec
/usr/bin/ruby3.1 -I/home/pravi/forge/ruby-team/ruby-invisible-captcha/vendor/bundle/ruby/3.1.0/gems/rspec-core-3.12.1/lib:/home/pravi/forge/ruby-team/ruby-invisible-captcha/vendor/bundle/ruby/3.1.0/gems/rspec-support-3.12.0/lib /home/pravi/forge/ruby-team/ruby-invisible-captcha/vendor/bundle/ruby/3.1.0/gems/rspec-core-3.12.1/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb

Randomized with seed 748

InvisibleCaptcha::ControllerExt
  without invisible_captcha_timestamp in session
    passes if disabled at app level
    passes if disabled at action level
    fails like if it was submitted too fast
  submission timestamp_threshold
    allows a new timestamp to be set in the on_timestamp_spam callback
    allows a custom on_timestamp_spam callback
    fails if submission before timestamp_threshold
    successful submissions
      allow to set a custom timestamp_threshold per action
      passes if submission on or after timestamp_threshold
  spinner attribute
    fails with no spam, but mismatch of spinner
    passes with no spam and spinner match
  honeypot attribute
    honeypot is removed from params if you use a custom honeypot
    allow a custom on_spam callback
    fails with spam
    passes with no spam
    ActiveSupport::Notifications
      dispatches an `invisible_captcha.spam_detected` event (FAILED - 1)

InvisibleCaptcha::ViewHelpers
  should set spam timestamp
  with CSP nonce
  with specific honeypot
  with custom html options
  generated html + styles
  with no arguments
  with specific honeypot and scope
  injectable_styles option
    if injectable_styles is set, do not append styles inline
    by default, render styles along with the honeypot
  should have spinner field
    that does not exist if spinner_enabled is false
    that exists by default, spinner_enabled is true
  honeypot visibilty
    visible from given instance (default override)
    invisible from given instance (default override)
    visible from defaults

InvisibleCaptcha
  initialize with defaults
  allow setup via block
  It uses I18n when available

Failures:

  1) InvisibleCaptcha::ControllerExt honeypot attribute ActiveSupport::Notifications dispatches an `invisible_captcha.spam_detected` event
     Failure/Error:
       expect(dummy_handler).to receive(:handle_event).once.with(
         message: "Invisible Captcha honeypot param 'subtitle' was present.",
         remote_ip: '0.0.0.0',
         user_agent: 'Rails Testing',
         controller: 'topics',
         action: 'create',
         url: 'http://test.host/topics',
         params: {
           topic: { subtitle: "foo"},
           controller: 'topics',
     
       #<Double (anonymous)> received :handle_event with unexpected arguments
         expected: ({:action=>"create", :controller=>"topics", :message=>"Invisible Captcha honeypot param 'subtitle' was...itle=>"foo"}}, :remote_ip=>"0.0.0.0", :url=>"http://test.host/topics", :user_agent=>"Rails Testing"}) (keyword arguments)
              got: ({:action=>"create", :controller=>"topics", :message=>"Invisible Captcha honeypot param 'subtitle' was...tle"=>"foo"}}, :remote_ip=>"0.0.0.0", :url=>"http://test.host/topics", :user_agent=>"Rails Testing"}) (options hash)
       Diff:
       @@ -2,7 +2,7 @@
          :controller=>"topics",
          :message=>"Invisible Captcha honeypot param 'subtitle' was present.",
          :params=>
       -   {:action=>"create", :controller=>"topics", :topic=>{:subtitle=>"foo"}},
       +   {"action"=>"create", "controller"=>"topics", "topic"=>{"subtitle"=>"foo"}},
          :remote_ip=>"0.0.0.0",
          :url=>"http://test.host/topics",
          :user_agent=>"Rails Testing"}]
       
     # ./spec/controllers_spec.rb:151:in `block (4 levels) in <top (required)>'

Finished in 10.29 seconds (files took 0.94029 seconds to load)
32 examples, 1 failure

Failed examples:

rspec ./spec/controllers_spec.rb:150 # InvisibleCaptcha::ControllerExt honeypot attribute ActiveSupport::Notifications dispatches an `invisible_captcha.spam_detected` event

Randomized with seed 748

/usr/bin/ruby3.1 -I/home/pravi/forge/ruby-team/ruby-invisible-captcha/vendor/bundle/ruby/3.1.0/gems/rspec-core-3.12.1/lib:/home/pravi/forge/ruby-team/ruby-invisible-captcha/vendor/bundle/ruby/3.1.0/gems/rspec-support-3.12.0/lib /home/pravi/forge/ruby-team/ruby-invisible-captcha/vendor/bundle/ruby/3.1.0/gems/rspec-core-3.12.1/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb failed

```

**Comment by pravi** [comment]
@markets may be you are using an old Gemfile.lock ? Can you try running it again with a fresh Gemfile.lock and share your Gemfile.lock to compare with mine.

**Comment by pravi** [comment]
It is also failing in debian's salsa ci https://salsa.debian.org/ruby-team/ruby-invisible-captcha/-/jobs/3785616

**Comment by markets** [comment]
This is not the official repo... It seems an outdated fork. Could you please try with this repo hosted on GitHub under my account?

**Comment by pravi** [comment]
I shared the salsa ci link just to show the issue is reproducible in a clean Debian unstable system and not specific to my local machine. We maintain 1400+ ruby gems in Debian (gems that are used by web applications like redmine and gitlab or tools like jekyll so these can be installed with apt package manager without needing any ruby specific knowledge). We make sure the gem works in Debian environment by running tests during build. Sometimes the issue may be specific to Debian, if you can share your Gemfile.lock (many projects include it in their repos), we can confirm whether the issue is specific to Debian environment or not. Last ci run was 3 weeks ago, may be some gem versions changed after ci passed. If you can manually trigger the ci run now (assuming it will generate a new Gemfile.lock), that will also help.

**Comment by markets** [comment]
But again, this repo you sent me (https://salsa.debian.org/ruby-team/ruby-invisible-captcha) seems an outdated fork of mine, it doesn't include latest changes in the canonical repo. It even contains a `.travis.yml` file wich was removed here like 2 years ago (https://github.com/markets/invisible_captcha/pull/104) ...

You can try yourself to run tests using the official repo (this one) and the official docs to run the test suite (`bundle exec rspec`). You'll see they are correct.

On the other hand, including the `Gemfile.lock` is a bad practice for gems: https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/

**Comment by pravi** [comment]
This repo is generated from https://rubygems.org/gems/invisible_captcha/versions/2.0.0. So this should match with 2.0.0 tag on your repo. We usually update our repo when there is a new release, so we are interested only in tags not the whole repo. You can see the .gem file on rubygems.org still contains the .travis.yml file. The command difference does not matter here, as bundle exec rspec also fails the same way. May be the .gem file is missing some files present in the repo? I can try to switch to 2.0.0 tag from this repo instead of the .gem from rubygems.org. I will report here how that goes.

```
pravi@mahishi:/tmp$ wget https://rubygems.org/downloads/invisible_captcha-2.0.0.gem
--2023-03-04 00:16:59--  https://rubygems.org/downloads/invisible_captcha-2.0.0.gem
Resolving rubygems.org (rubygems.org)... 151.101.193.227, 151.101.65.227, 151.101.1.227, ...
Connecting to rubygems.org (rubygems.org)|151.101.193.227|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 26112 (26K) [application/octet-stream]
Saving to: ‘invisible_captcha-2.0.0.gem’

invisible_captcha-2.0. 100%[==========================>]  25.50K  --.-KB/s    in 0.03s   

2023-03-04 00:17:00 (910 KB/s) - ‘invisible_captcha-2.0.0.gem’ saved [26112/26112]

pravi@mahishi:/tmp$ gem2tgz invisible_captcha-2.0.0.gem 
tar xfm /tmp/invisible_captcha-2.0.0.gem
"tar xzfm data.tar.gz"
tar czf /tmp/invisible-captcha-2.0.0.tar.gz invisible-captcha-2.0.0
pravi@mahishi:/tmp$ tar -zxvf invisible-captcha-2.0.0.tar.gz 
invisible-captcha-2.0.0/
invisible-captcha-2.0.0/invisible_captcha.gemspec
invisible-captcha-2.0.0/.gitignore
invisible-captcha-2.0.0/Rakefile
invisible-captcha-2.0.0/spec/
invisible-captcha-2.0.0/spec/view_helpers_spec.rb
invisible-captcha-2.0.0/spec/dummy/
invisible-captcha-2.0.0/spec/dummy/log/
invisible-captcha-2.0.0/spec/dummy/log/.gitkeep
invisible-captcha-2.0.0/spec/dummy/Rakefile
invisible-captcha-2.0.0/spec/dummy/config/
invisible-captcha-2.0.0/spec/dummy/config/application.rb
invisible-captcha-2.0.0/spec/dummy/config/locales/
invisible-captcha-2.0.0/spec/dummy/config/locales/en.yml
invisible-captcha-2.0.0/spec/dummy/config/boot.rb
invisible-captcha-2.0.0/spec/dummy/config/environment.rb
invisible-captcha-2.0.0/spec/dummy/config/routes.rb
invisible-captcha-2.0.0/spec/dummy/config/secrets.yml
invisible-captcha-2.0.0/spec/dummy/config/environments/
invisible-captcha-2.0.0/spec/dummy/config/environments/production.rb
invisible-captcha-2.0.0/spec/dummy/config/environments/test.rb
invisible-captcha-2.0.0/spec/dummy/config/environments/development.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/
invisible-captcha-2.0.0/spec/dummy/config/initializers/backtrace_silencers.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/filter_parameter_logging.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/session_store.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/cookies_serializer.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/mime_types.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/wrap_parameters.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/secret_token.rb
invisible-captcha-2.0.0/spec/dummy/config/initializers/inflections.rb
invisible-captcha-2.0.0/spec/dummy/bin/
invisible-captcha-2.0.0/spec/dummy/bin/bundle
invisible-captcha-2.0.0/spec/dummy/bin/setup
invisible-captcha-2.0.0/spec/dummy/bin/rails
invisible-captcha-2.0.0/spec/dummy/bin/rake
invisible-captcha-2.0.0/spec/dummy/lib/
invisible-captcha-2.0.0/spec/dummy/lib/assets/
invisible-captcha-2.0.0/spec/dummy/lib/assets/.gitkeep
invisible-captcha-2.0.0/spec/dummy/README.md
invisible-captcha-2.0.0/spec/dummy/app/
invisible-captcha-2.0.0/spec/dummy/app/controllers/
invisible-captcha-2.0.0/spec/dummy/app/controllers/topics_controller.rb
invisible-captcha-2.0.0/spec/dummy/app/controllers/application_controller.rb
invisible-captcha-2.0.0/spec/dummy/app/models/
invisible-captcha-2.0.0/spec/dummy/app/models/topic.rb
invisible-captcha-2.0.0/spec/dummy/app/mailers/
invisible-captcha-2.0.0/spec/dummy/app/mailers/.gitkeep
invisible-captcha-2.0.0/spec/dummy/app/views/
invisible-captcha-2.0.0/spec/dummy/app/views/topics/
invisible-captcha-2.0.0/spec/dummy/app/views/topics/new.html.erb
invisible-captcha-2.0.0/spec/dummy/app/views/layouts/
invisible-captcha-2.0.0/spec/dummy/app/views/layouts/application.html.erb
invisible-captcha-2.0.0/spec/dummy/app/assets/
invisible-captcha-2.0.0/spec/dummy/app/assets/config/
invisible-captcha-2.0.0/spec/dummy/app/assets/config/manifest.js
invisible-captcha-2.0.0/spec/dummy/app/assets/javascripts/
invisible-captcha-2.0.0/spec/dummy/app/assets/javascripts/application.js
invisible-captcha-2.0.0/spec/dummy/app/assets/stylesheets/
invisible-captcha-2.0.0/spec/dummy/app/assets/stylesheets/application.css
invisible-captcha-2.0.0/spec/dummy/app/helpers/
invisible-captcha-2.0.0/spec/dummy/app/helpers/application_helper.rb
invisible-captcha-2.0.0/spec/dummy/config.ru
invisible-captcha-2.0.0/spec/dummy/public/
invisible-captcha-2.0.0/spec/dummy/public/500.html
invisible-captcha-2.0.0/spec/dummy/public/favicon.ico
invisible-captcha-2.0.0/spec/dummy/public/404.html
invisible-captcha-2.0.0/spec/dummy/public/422.html
invisible-captcha-2.0.0/spec/controllers_spec.rb
invisible-captcha-2.0.0/spec/invisible_captcha_spec.rb
invisible-captcha-2.0.0/spec/spec_helper.rb
invisible-captcha-2.0.0/CHANGELOG.md
invisible-captcha-2.0.0/lib/
invisible-captcha-2.0.0/lib/invisible_captcha/
invisible-captcha-2.0.0/lib/invisible_captcha/version.rb
invisible-captcha-2.0.0/lib/invisible_captcha/railtie.rb
invisible-captcha-2.0.0/lib/invisible_captcha/view_helpers.rb
invisible-captcha-2.0.0/lib/invisible_captcha/form_helpers.rb
invisible-captcha-2.0.0/lib/invisible_captcha/controller_ext.rb
invisible-captcha-2.0.0/lib/invisible_captcha.rb
invisible-captcha-2.0.0/README.md
invisible-captcha-2.0.0/gemfiles/
invisible-captcha-2.0.0/gemfiles/rails_6.0.gemfile
invisible-captcha-2.0.0/gemfiles/rails_6.1.gemfile
invisible-captcha-2.0.0/gemfiles/rails_5.2.gemfile
invisible-captcha-2.0.0/.travis.yml
invisible-captcha-2.0.0/Appraisals
invisible-captcha-2.0.0/Gemfile
invisible-captcha-2.0.0/LICENSE
invisible-captcha-2.0.0/.rspec
```

**Comment by pravi** [comment]
Sorry for missing the detail that, I was working with 2.0.0 tag. After taking a snapshot of the current git master, the tests are passing. If you can make a new release, that would be nice as we could then ship the same version used by gitlab (gitlab is still using 2.0.0 version) and also have the tests passing.

**Comment by markets** [comment]
> Sorry for missing the detail that

No problem @pravi!

> If you can make a new release, that would be nice as we could then ship the same version used by gitlab

Sorry but this is not how I understand open-source. I'll cut a new release when I think it's ready and it makes sense for the final users. I don't want to generate noise for users updating the gem, reviewing the changelog, testing their apps, ... just because you want to pass all tests of a code released ~2 years ago (Ruby 3.1 didn't even exist at those times, neither Rails 7). Hopefully you'll understand that! But if you need some special/custom requests, you can sponsor my open-source work to get higher priority: https://github.com/sponsors/markets.

**Comment by pravi** [comment]
If the changes are only for tests and not to the main code, then we don't need a new release. When tests fail there is usually two possibilities, either the code needs changing or the tests need changes. Since I'm able to fix the failures from snapshot of master, that will be enough for Debian. New release request is a wishlist entry just to be on the safer side. I will close this now. Thanks a lot for the help in finding a fix.
